cmd/snap/cmd_run.go

```diff
@@ -448,13 +448,26 @@ func (x *cmdRun) snapRunTimer(snapApp, timer string, args []string) error {
 
 var osReadlink = os.Readlink
 
-func isReexeced() bool {
+// snapdHelperPath return the path of a helper like "snap-confine" or
+// "snap-exec" based on if snapd is re-execed or not
+func snapdHelperPath(toolName string) (string, error) {
 	exe, err := osReadlink("/proc/self/exe")
 	if err != nil {
-		logger.Noticef("cannot read /proc/self/exe: %v", err)
-		return false
+		return "", fmt.Errorf("cannot read /proc/self/exe: %v", err)
 	}
-	return strings.HasPrefix(exe, dirs.SnapMountDir)
+	// no re-exec
+	if !strings.HasPrefix(exe, dirs.SnapMountDir) {
+		return filepath.Join(dirs.DistroLibExecDir, toolName), nil
+	}
+
+	// snapBase will be "/snap/{core,snapd}/$rev/" because
+	// the snap binary is always at $root/usr/bin/snap
+	snapBase := filepath.Clean(filepath.Join(filepath.Dir(exe), "..", ".."))
```

Comment:
Can we please check that the assumption you spell out holds?

```
// The logic below only works if the last two path components are /usr/bin
if !strings.HasSuffix(exe, "/usr/bin/" + filepath.Base(exe)) {
    return filepath.Join(dirs.DistroLibExecDir, toolName), nil
}
```

Reply:
Hm, we shold also probably warn about this, yes?